app/api/Files.scala

```diff
@@ -523,14 +523,47 @@ class Files @Inject()(
   /**
     * Upload a file to a specific dataset
     */
-  def uploadToDataset(dataset_id: UUID, showPreviews: String = "DatasetLevel", originalZipFile: String = "", flagsFromPrevious: String = "", extract: Boolean = true) = PermissionAction(Permission.AddResourceToDataset, Some(ResourceRef(ResourceRef.dataset, dataset_id)))(parse.multipartFormData) { implicit request =>
+  def uploadToDataset(dataset_id: UUID, showPreviews: String = "DatasetLevel", originalZipFile: String = "", flagsFromPrevious: String = "", extract: Boolean = true, folder_id: Option[String] = None) = PermissionAction(Permission.AddResourceToDataset, Some(ResourceRef(ResourceRef.dataset, dataset_id)))(parse.multipartFormData) { implicit request =>
     datasets.get(dataset_id) match {
       case Some(dataset) => {
-        val uploadedFiles = FileUtils.uploadFilesMultipart(request, Some(dataset), showPreviews = showPreviews, originalZipFile = originalZipFile, flagsFromPrevious = flagsFromPrevious, runExtractors = extract, apiKey = request.apiKey)
-        uploadedFiles.length match {
-          case 0 => BadRequest("No files uploaded")
-          case 1 => Ok(Json.obj("id" -> uploadedFiles.head.id))
-          case _ => Ok(Json.obj("ids" -> uploadedFiles.toList))
+        folder_id match {
```

Comment:
I think we can simplify (not tested)
```
val folder = folder_id.flatmap(x => folders.get(x))
val uploadedFiles = FileUtils.uploadFilesMultipart(request, Some(dataset), folder, showPreviews = showPreviews, originalZipFile = originalZipFile, flagsFromPrevious = flagsFromPrevious, runExtractors = extract, apiKey = request.apiKey) 
uploadedFiles.length match {
    case 0 => BadRequest("No files uploaded")
    case 1 => Ok(Json.obj("id" -> uploadedFiles.head.id))
    case _ => Ok(Json.obj("ids" -> uploadedFiles.toList))
}
```
This will use flatmap (if there is a value, return an option) to get the folder and then just call the next function with ether None, or Some(folder).

Reply:
This works.

However, I have another problem.

I have been testing this with the following python script 

https://github.com/tcnichol/scripts/blob/master/upload_with_folder.py

The method 'upload_a_file_to_dataset_with_folder' works. 

Before I added the folder_id parameter, the method 'upload_a_file_to_dataset' worked. Now it no longer does; I get a 500 error unless I supply a value for folder_id.

But since folder_id is not a required parameter and has a default value, I am not sure why this is happening. Obviously before the same type of call worked but without any optional parameters being supplied. 

I am going to try to add the folder_id to the multipart form data.

Reply:
I am now using a conditional that seems to handle all cases without breaking. 

Regrettably it kept giving me an error if I tried to put an Option[UUID] in the api route and didn't send it, but this seems to match what is done in a few other cases with optional uuid parameters.